[Malcolm Ang] iP

malcolmang opened the pull request.

malcolmang committed:
Add Increment: Level-0

Added Level-0 functionality with the "Dude" chatbot.

malcolmang committed:
Edit class names to be consistent as instructed

malcolmang committed:
Add classes to help with formatting, indentation etc.

Added a "Display" class to manage the entire display of the User Interface, and a "Dialog" class to generate a text box and indentation between linebreaks

malcolmang committed:
Add Increment: Level 1

Program greets the user, echos commands until "bye" is given.

malcolmang committed:
Fix minor bugs relating to dialog box enforcement

The dialog box should be more robust and able to take in words longer than the max line space.

malcolmang committed:
Add Increment: Level-2

Program now can add to a list of tasks to do.

malcolmang committed:
Add dialog extension function

Allows for dialog box to take in more items than the height of the ASCII art. List can now support up to 100 items. Additional minor bug fixes for indentation

malcolmang committed:
Add Increment: Level 3, A-Classes

Added the ability to mark and unmark tasks. Added the "ToDoItem" class to contain all the information each task can provide.

malcolmang committed:
Edit ToDoItem class

Made sure that the constructor initializes "isDone" as false

malcolmang committed:
Format code to adhere with coding standard

malcolmang committed:
Refactor indent function into multiple parts

Made it easier to read and parse. Also should be the most robust indentation possible.

malcolmang committed:
Refactor and document code

Code is now documented with Javadoc formatting and refactored for readability. Made minor changes to change methods to static where possible, and reduced the amount of "magic numbers".

malcolmang committed:
Add Increment: Level 4

Program can now track deadlines, events and standard task items.
Refactored names and variables to be more consistent and
packaged classes into a more organized structure.

malcolmang committed:
Fix markdown

Minor fix to the CONTRIBUTORS.md file
which appears to have gotten accidentally renamed
from auto-refactoring.

malcolmang committed:
Make minor change to print() function

A small change to add a blank line to the top
of the print function in UserInterface class.

eehongchan commented on src/main/java/tasks/tasktypes/ToDoTask.java:
```diff
@@ -0,0 +1,21 @@
+     * Provides type icon to be used in string formatting.
+     * @return Type icon "T" for "ToDo".
+     */
+    @Override
+    protected String getTypeIcon() {
+        return "T";
+    }
```
Good use of overriding!

eehongchan reviewed (commented):
Excellent coding standard!

HT-T commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,161 @@
+import UI.UserInterface;
+import tasks.tasktypes.DeadlineTask;
+import tasks.tasktypes.EventTask;
+import tasks.tasktypes.ToDoTask;
+import tasks.TaskList;
```
I like how you adhere to the guidelines: only import packages you need

HT-T commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,161 @@
+    public static final String INPUT_DEADLINE = "deadline";
+    public static final String INPUT_EVENT = "event";
+
+    /** tasks.TaskList class that contains all items */
```
Nice to see comments that helps reviewers to understand the code

HT-T commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,161 @@
-                + "| |_| | |_| |   <  __/\n"
-                + "|____/ \\__,_|_|\\_\\___|\n";
-        System.out.println("Hello from\n" + logo);
+        greet();
+        programBody();
+        exit();
```
I like how the SLAP principle is used here. Clean and readable.

HT-T commented on src/main/java/tasks/tasktypes/ToDoTask.java:
```diff
@@ -0,0 +1,21 @@
+ */
+public class ToDoTask extends Task {
+    public ToDoTask(String text) {
+        super(text);
```
Good implementation of OOP overall

HT-T commented on src/main/java/UI/DialogBox.java:
```diff
@@ -0,0 +1,215 @@
+            // string is too long, then add entire buffer string
+            // as a row and repeat.
+            for (String word : item.split(" ")) {
+                if (word.length() > maxLength) {
```
I like this well-thought functionality

HT-T commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,161 @@
+    public static void programBody() {
+        while (true) {
+            String input = UserInterface.getInput();
+            if (isTerminatingInput(input)) {
```
Perhaps you could consider switch-case? I have seen other coder who used switch-case for better readability. 
However this is a very minor issue.

HT-T reviewed (commented):
Overall the code is up to standard and implemented correctly. Looks good.

eehongchan commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,161 @@
+    private static final String INPUT_MARK = "mark";
+    private static final String INPUT_UNMARK = "unmark";
+    public static final String INPUT_DEADLINE = "deadline";
+    public static final String INPUT_EVENT = "event";
```
Good use of final variables!

eehongchan commented on src/main/java/UI/DialogBox.java:
```diff
@@ -0,0 +1,215 @@
+                if (bufferString.length() + word.length() > maxLength) {
+                    // if length of buffer string + new word exceeds max length
+                    // add buffer string as a new row and reset buffer string
+                    textRowsIndividual.add(generateIndent(indentLength, INDENT_CHARACTER) + bufferString.trim());
```
Can consider breaking up into 2 lines

eehongchan commented on src/main/java/UI/DialogBox.java:
```diff
@@ -0,0 +1,215 @@
+        // Vertical lines at the side of the text
+        for (String row : textRows) {
+            textBoxRows.add(" " + LINE_VERTICAL_CHARACTER + generateIndent(HORIZONTAL_PADDING, INDENT_CHARACTER)
+                    + row + generateIndent(width - row.length() - HORIZONTAL_PADDING, ' ') + LINE_VERTICAL_CHARACTER);
```
Can consider splitting into 2 lines

eehongchan reviewed (commented):
Good code, but could split up lengthy lines

malcolmang committed:
Edit documentation

Just minor formatting changes to javadocs

malcolmang committed:
Add error handling (incomplete)

Added various exceptions to be thrown when input is being validated.
InputValidator class added to encapsulate these methods as well.

malcolmang committed:
Add more exceptions

Made program more robust

malcolmang committed:
Add Increment: Level-5

malcolmang committed:
Merge branch 'branch-Level-5'

malcolmang committed:
Rename CustomException to DukeException

Minor refactor for readability

malcolmang committed:
Document code

Slight refactoring for variable/method readability. Also
added documentation to adhere to code standard

malcolmang committed:
Refactor files

Another minor refactor

malcolmang committed:
Add Increment: Level 6

Items can now be deleted from the list.
Minor refactoring to move message formatting
to a separate class.

malcolmang committed:
Add Increment: Level 7

Duke now reads from and writes to a persistent save file.

malcolmang committed:
Merge branch 'branch-Level-6'

malcolmang committed:
Merge branch 'branch-Level-7'

Resolved:
# Conflicts:
#	src/main/java/duke/Duke.java

malcolmang committed:
Refactor and enforce code style

Just a minor auto-reformatting

malcolmang committed:
Change .gitignore to ignore MANIFEST.MF files

This is not needed for the .jar to run

malcolmang committed:
Fix bug with saving and loading files

As a .jar the status icon was not recognised and written to,
resulting in incorrect loads. That was fixed and a working jar is created.

malcolmang committed:
Fix output string formatting

exetr commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,236 @@
+     */
+    private static int extractNumber(String input) throws NotAnIntegerException {
+        String number = input.split(" ")[1];
+        String command = input.split(" ")[0];
```
Perhaps you can consider replacing the delimiter here with a variable instead, so as to eliminate magic literals.

exetr commented on src/main/java/duke/error/exceptions/ItemNotFoundException.java:
```diff
@@ -0,0 +1,37 @@
+            tempString += "\nRemember that your to-do list is 1-indexed (0 is not valid).";
+        }
+        tempString += (Duke.TASK_LIST.getItemCount() == 0) ? "\nYour to-do list is currently empty." : String.format(
+                "\nYour to-do list currently has %s item(s).", Duke.TASK_LIST.getItemCount());
```
While this implementation is perfectly functional, perhaps you may want to consider splitting it up into an if-else statement, so as to make this particular line is more readable to others.

exetr commented on src/main/java/duke/io/FileManager.java:
```diff
@@ -0,0 +1,165 @@
+    private static final String SEPARATOR = "|";
+
+    /** Boolean flag that triggers if file was loaded. */
+    private static boolean wasLoaded = false;
```
Perhaps you can consider changing this variable name to `isLoaded` to maintain consistency with the coding standard.

exetr commented on src/main/java/duke/io/FileManager.java:
```diff
@@ -0,0 +1,165 @@
+     *
+     * @return {@link TaskList} instance, empty if no file exists
+     */
+    public static TaskList LoadTaskList() {
```
You could consider renaming this function to `loadTaskList` as per the coding standard, where function names should be in camelCase.

exetr commented on src/main/java/duke/tasks/tasktypes/DeadlineTask.java:
```diff
@@ -0,0 +1,97 @@
+        String subcommand = sections[3];
+        String text = sections[2];
+        String icon = sections[1];
+        DeadlineTask bufferTask = new DeadlineTask(text + "/by" + subcommand);
```
Since you have already defined the `/by` above as a final variable, you can consider replacing the magic string here.

exetr commented on src/main/java/duke/tasks/tasktypes/EventTask.java:
```diff
@@ -0,0 +1,99 @@
+        String subcommand = sections[3];
+        String text = sections[2];
+        String icon = sections[1];
+        EventTask bufferTask = new EventTask(text + "/at" + subcommand);
```
Similarly, since the `/at` final is already declared above, you can consider replacing the magic string here.

exetr commented on src/main/java/duke/ui/StringFormatting.java:
```diff
@@ -0,0 +1,84 @@
+     * @return formatted string
+     */
+    public static String formatNumberOfTasksString(int numberOfTasks) {
+        return "\nThere " + (numberOfTasks == 1 ? "is" : "are")
+                + " currently " + numberOfTasks + " task"
+                + (numberOfTasks == 1 ? "" : "s") + " in your list.";
+    }
```
Similar to previously, while functionally this is valid, you may want to consider splitting this logic up with the use of if-else statements, so as to make this section more reader-friendly.

exetr reviewed (commented):
Great job on your iP so far. I have left some comments and suggestions for you to consider.

malcolmang committed:
Refactor class names, fix code style violations

Edited class names to reflect requirements of A-MoreOOP,
and refactored to more closely adhere to code standard

malcolmang committed:
Add Increment: Level 9

Program can now do a simple case-insensitive search for tasks containing
a specified substring with command "find".

malcolmang committed:
Add javadoc comments

Added more javadoc comments to be fully compliant with the coding
standard.

malcolmang committed:
Merge pull request #1 from malcolmang/branch-Level-9

Add Increment: Level 9

malcolmang committed:
Merge branch 'master' into branch-A-JavaDoc

# Conflicts:
#	src/main/java/duke/Duke.java

malcolmang committed:
Merge pull request #2 from malcolmang/branch-A-JavaDoc

Add javadoc comments

malcolmang committed:
Refactor code to comply with coding standard

Import every package individually and add javadoc for every method

malcolmang committed:
Fix minor typos

malcolmang committed:
Refactor to be consistent with coding standard

Reformatted try-catch and if-else blocks to adhere with
coding standard

malcolmang committed:
Add markdown documentation

For github pages functionality
